File: client-android/serenada-core/src/main/java/app/serenada/core/SerenadaLogger.kt
package app.serenada.core

import android.util.Log

enum class SerenadaLogLevel { DEBUG, INFO, WARNING, ERROR }

interface SerenadaLogger {
    fun log(level: SerenadaLogLevel, tag: String, message: String)
}

class AndroidSerenadaLogger : SerenadaLogger {
    override fun log(level: SerenadaLogLevel, tag: String, message: String) {
        when (level) {
            SerenadaLogLevel.DEBUG -> Log.d(tag, message)
            SerenadaLogLevel.INFO -> Log.i(tag, message)
            SerenadaLogLevel.WARNING -> Log.w(tag, message)
            SerenadaLogLevel.ERROR -> Log.e(tag, message)
        }
    }
}

File: client-ios/SerenadaCore/Sources/SerenadaLogger.swift
import Foundation

public enum SerenadaLogLevel: Int, Comparable, Sendable {
    case debug = 0, info, warning, error

    public static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
}

public protocol SerenadaLogger: AnyObject, Sendable {
    func log(_ level: SerenadaLogLevel, tag: String, _ message: String)
}

public final class PrintSerenadaLogger: SerenadaLogger {
    public init() {}

    public func log(_ level: SerenadaLogLevel, tag: String, _ message: String) {
        let levelLabel: String
        switch level {
        case .debug: levelLabel = "DEBUG"
        case .info: levelLabel = "INFO"
        case .warning: levelLabel = "WARN"
        case .error: levelLabel = "ERROR"
        }
        print("[\(levelLabel)] [\(tag)] \(message)")
    }
}
